src/metabase/driver.clj

```diff
@@ -1,232 +1,411 @@
+
+    (the-driver :postgres) ; -> :postgres
+    (the-driver :baby)     ; -> Exception"
+  [driver]
+  (when-not (keyword? driver)
+    (throw (Exception. (str (tru "Invalid driver name: {0}" driver)))))
+  (load-driver-namespace-if-needed driver)
+  driver)
+
+(defn- check-abstractness-hasnt-changed
+  "Check to make sure we're not trying to change the abstractness of an already registered driver"
+  [driver new-abstract?]
+  (let [old-abstract? (abstract? driver)]
+    (when (and (registered? driver) (not= (boolean old-abstract?) (boolean new-abstract?)))
+      (throw (Exception. (str (tru "Error: attempting to change {0} property `:abstract?` from {1} to {2}."
+                                   driver old-abstract? new-abstract?)))))))
+
+(defn add-parent!
+  "Add a new parent to `driver`."
+  [driver new-parent]
+  (load-driver-namespace-if-needed driver)
+  (load-driver-namespace-if-needed new-parent)
+  (alter-var-root #'hierarchy derive driver new-parent))
+
+(defn register!
+  "Register a driver.
+
+    (register! :sql, :abstract? true)
+
+    (register! :postgres, :parent :sql-jdbc)
+
+  Valid options are:
+
+  ###### `:parent` (default = none)
+
+  Parent driver(s) to derive from. Drivers inherit method implementations from their parents similar to the way
+  inheritance works in OOP. Specify multiple direct parents by passing a collection of parents.
+
+  You can add additional parents to a driver using `add-parent!` below; this is how test extensions are implemented.
+
+  ###### `:abstract?` (default = false)
+
+  Is this an abstract driver (i.e. should we hide it in the admin interface, and disallow running queries with it)?
+
+  Note that because concreteness is implemented as part of our keyword hierarchy it is not currently possible to
+  create an abstract driver with a concrete driver as its parent, since it would still ultimately derive from
+  `::concrete`."
+  [driver & {:keys [parent abstract?]}]
+  {:pre [(keyword? driver)]}
+  ;; validate that the registration isn't stomping on things
+  (check-abstractness-hasnt-changed driver abstract?)
+  ;; ok, if that was successful we can derive the driver from `::driver`/`::concrete` and parent(s)
+  (let [derive! (partial alter-var-root #'hierarchy derive driver)]
+    (derive! ::driver)
+    (when-not abstract?
+      (derive! ::concrete))
+    (doseq [parent (cond
+                     (coll? parent) parent
+                     parent         [parent])
+            :when  parent]
+      (load-driver-namespace-if-needed parent)
+      (derive! parent)))
+  ;; ok, log our great success
+  (when-not (metabase.driver/abstract? driver)
+    (log/info (trs "Registered driver {0} {1}" (u/format-color 'blue driver) (u/emoji "🚚")))))
+
+(defn dispatch-on-driver
+  "Dispatch function to use for driver multimethods. Dispatches on first arg, a driver keyword; loads that driver's
+  namespace if not already done."
+  [x & _]
+  (the-driver x))
```

Comment:
Having side-effects in the dispatch fn scares me, although as far as I can see, this should work as indented and won't get fouled by caching or some such